### PR TITLE
Add runcron admin management command

### LINKED ISSUE #31: Extend runcron management command
This task is to extend the runcron command so that it reads from a "scheduler" model and does the tasks as specified when manage.py runcron process_items is called. 

The schdeuler model would have 3 fields, one field would be the "command" to run, the other would be the "data", the last would be a "success" field. Storing the items in the data field in the model in json would probably be easiest. 

For example, the model would include things such as "send_email" which would take the "data" field and call the sendemail function. 

You will also need to add arguments to runcron for "build_items" and "process_items" as well as when ran without arguments it would run build_items then process_items in order.

### PR DESCRIPTION
**Pull request by sounak98**
This PR will add the `runcron` admin management command in django to run scheduled tasks.

Fixes #31.

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Run command `python manage.py runcron <command> --num_workers=5 --timeout=10`

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by sounak98** [comment]
@botanicvelious check the structure of the management command, does it look good? I will now proceed with creating the `Scheduler` model.

**Comment by botanicvelious** [comment]
general idea looks good, the 2 commands should be "build_items" and "process_items" however. Also the -t and -n only would be arguments for "process_items" so that would also be good to change or explicitly ignore. Lastly should set it as a default to run build then process if neither one is specified.

**Comment by sounak98** [comment]
@botanicvelious have updated the structure to allow the specified commands `build_items` and `process_items` and take the former as the default one when nothing is mentioned.

**Review by botanicvelious — CHANGES_REQUESTED** [review]
It would be a good idea to use a TextField rather than a char for the data column, as 1000 chars could likely not be enough for some items so just use a TextField so you never have to worry about it.

Also, it should build, then run not just build if there are no additional arguments.

Lastly please dont commit changes to the .db files, that way we dont run into conflicts. You can set this by .git/info/exclude and git update-index --skip-worktree [<file>...]

**Comment by sounak98** [comment]
- [x] `data` field is updated to `TextField` from `CharField` (was already going to do this! 😅)
- [x] the `db` files have been removed from the commits
- [x] the management command `runcron` supports two subcommands, viz. `process_items` and `build_items`. The `process_items` takes up the unattended scheduled tasks and runs them on a thread pool before marking them successful.

The subcommands of the `runcron` management command are:
- `python manage.py runcron process_items --num_workers=5 --timeout=10`
- `python manage.py runcron build_items`

@botanicvelious I have left the util functions `send_email` and `send_irc_msg` empty, and also the `build_items` function; all of which can be worked on in different PRs.

**Comment by botanicvelious** [comment]
Overall looks good, you missed that 

```
manage.py runcron
Build Items
```

Should build then run, not just build. So the output would be 

```
manage.py runcron
Build Items
No scheduled tasks
```

Also, the code looks to not use the timeout on a worker process at all. You could easily test your process timeout function by waiting an amount of time greater than the timeout in for example send_email and waiting a bit less in the send_irc_msg.

Lastly, the success doesn't seem to take into account if the processed item was actually successful, just looks to assume it was. Thinking about it would also be useful to have a "failure_message" field for when there is a failure so that can be logged or notify someone etc. Additionally, it looks like the default is failed, when the default should be null, then updated to success or failure based on if the process completes successfully.

**Comment by sounak98** [comment]
@botanicvelious `scheduler.success == True` means that the scheduled command has been successfully executed. In all other scenarios `scheduler.success == False`. I have added a log for the errors caused while executing the commands (if any). If the execution fails `scheduler.success` remains `False`.

My bad, I forgot about the timeout. I've added it.

**Comment by botanicvelious** [comment]
The issue with having is starting false and stay false is that you then have no way of knowing if it ran and failed as it says false in either case. That is why it's typically best to have a null that is changed to the proper value. This way you can run the "process_items" against anything where status = null, then you know if something failed it is false if it succeeded its true, and you can easily have a process in "build_items" to make it so that it re-adds anything that is failed lets say up to 3 times or whatever. This way you have a full log of everything without having to rely on additional logging systems or otherwise.

If you want to get on irc I am online currently and we can chat on the best way to handle this.

**Comment by sounak98** [comment]
I agree with you @botanicvelious, I've updated the `Scheduler` to have a field for the last error and also `scheduler.success` is set to `None` when the `Scheduler` object is being created.

**Review by botanicvelious — COMMENTED** [review]
> Best practice in Django is to have a BooleanField accept only the two values: true and false. If you need to accept NULL values, you have to use a NullBooleanField.

Actually, my bad I wasn't thinking a default value. Everything looks good.

**Review by botanicvelious — APPROVED** [review]

**Comment by sounak98** [comment]
@botanicvelious django documentation says
> ### NullBooleanField
> Like BooleanField with null=True. Use that instead of this field as it’s likely to be deprecated in a future version of Django.

Oh, I just saw your comment.

**Comment by botanicvelious** [comment]
> 
> 
> @botanicvelious django documentation says
> 
> > ### NullBooleanField
> > Like BooleanField with null=True. Use that instead of this field as it’s likely to be deprecated in a future version of Django.

Yes, I realized after that I was not thinking.

File: gsoc/management/commands/runcron.py [hunk 1]
```diff
@@ -1,10 +1,82 @@
...
+
+        else:
+            self.stdout.write(self.style.ERROR('Command {}:{} failed with error: {}'
+                .format(scheduler.command, scheduler.id, err)), ending='\n')
+            scheduler.success = False
+            scheduler.last_error = err
+            scheduler.save()
+
+    def process_items(self, options):
+            schedulers = Scheduler.objects.exclude(success=True)
```
Comment (botanicvelious) [inline gsoc/management/commands/runcron.py]: This excludes true, but it needs to actually only include null

File: gsoc/management/commands/runcron.py [hunk 2]
```diff
@@ -1,10 +1,82 @@
...
     def handle(self, *args, **options):
-        # Add yout logic here
-        # This is the task that will be run
-        self.stdout.write('This was extremely simple!!!', ending='')
-        
\ No newline at end of file
+        if options['task']:
+            getattr(self, options['task'])(options)
+        else:
+            self.build_items(options)
```
Comment (botanicvelious) [inline gsoc/management/commands/runcron.py:79]: Shouldnt this be

```Python
self.build_items(options)
self.process_items(options)
```

File: gsoc/models.py [hunk 3]
```diff
@@ -3,6 +3,23 @@
...
+class Scheduler(models.Model):
+    commands = (
+        ('send_email', 'send_email'),
+        ('send_irc_msg', 'send_irc_msg')
+    )
+    
+    id = models.AutoField(primary_key=True)
+    command = models.CharField(name='command', max_length=20, choices=commands)
+    data = models.TextField(name='data')
+    success = models.BooleanField(name='success', null=True, default=None)
```
Comment (botanicvelious) [inline gsoc/models.py]: The default value of BooleanField is None when Field.default isn’t defined so is the default really necessary?

File: gsoc/models.py [hunk 4]
```diff
@@ -3,6 +3,23 @@
...
+    last_error = models.CharField(name='last_error', max_length=100, null=True, default=None)
```
Comment (botanicvelious) [inline gsoc/models.py]: A max_length=100 char field for errors may be quite small if there is like a stack trace or something especially. TextField probably the better choice.

**Review by botanicvelious — COMMENTED** [review]
Other then these items looks good.

File: gsoc/management/commands/runcron.py (hunk 2 above)
Comment (sounak98) [reply gsoc/management/commands/runcron.py:79]: `self.build_items(options)` run `self.process_items(options)` inside it.

File: gsoc/models.py (hunk 4 above)
Comment (sounak98) [reply gsoc/models.py]: Yeah, even I was thinking about this. I'll change this.

File: gsoc/management/commands/runcron.py (hunk 1 above)
Comment (sounak98) [reply gsoc/management/commands/runcron.py]: So, the processes that have been run before and have failed for some reason doesn't need to be run? Or should we add another flag for that? Something like `--exclude-failed`?

Comment (sounak98) [reply gsoc/management/commands/runcron.py]: And what should be the default behaviour?

Comment (botanicvelious) [reply gsoc/management/commands/runcron.py]: I would just never run failed stuff, we can always re-add anything that failed as part of the build_items

**Comment by sounak98** [comment]
@botanicvelious I have resolved all the reviews. Can you check?

File: gsoc/management/commands/runcron.py (hunk 2 above)
Comment (botanicvelious) [reply gsoc/management/commands/runcron.py:79]: That would then make it impossible to just build the items and not process them. That could be problematic down the line.

**Comment by botanicvelious** [comment]
@sounak98 looks good except for the one 

>`self.build_items(options)` run `self.process_items(options)` inside it.

if you can make that one last change i will merge the code

File: gsoc/management/commands/runcron.py (hunk 2 above)
Comment (sounak98) [reply gsoc/management/commands/runcron.py:79]: So then on running `runcron build_items` should it process it too?

I guess it shouldn't. Had a misconception.

**Comment by sounak98** [comment]
> if you can make that one last change i will merge the code

@botanicvelious Done!

File: gsoc/management/commands/runcron.py (hunk 2 above)
Comment (botanicvelious) [reply gsoc/management/commands/runcron.py:79]: >runcron build_items 

would only build

>runcron 

would build and run

**Comment by botanicvelious** [comment]
@sounak98 gsoc/management/commands/runcron.py is still unresolved?

File: gsoc/management/commands/runcron.py (hunk 2 above)
Comment (sounak98) [reply gsoc/management/commands/runcron.py:79]: The latest updates change the behaviour to exactly this.

**Comment by sounak98** [comment]
It's resolved @botanicvelious

**Comment by botanicvelious** [comment]
Merged!

**Pull request merged**